File: Task/Jordan-P-lya-numbers/Dart/jordan-p-lya-numbers.dart
import "dart:collection";
import "dart:io";

void main() {
  createJordanPolya();

  final belowHundredMillion = jordanPolyaSet.lastWhere((element) => element <= 100000000, orElse: () => null);
  List<int> jordanPolya = jordanPolyaSet.toList();

  print("The first 50 Jordan-Polya numbers:");
  for (int i = 0; i < 50; i++) {
    // Right-align each number in a 5-character wide space
    stdout.write(jordanPolya[i].toString().padLeft(5));
    if (i % 10 == 9) {
      print(""); // Newline every 10 numbers
    }
  }
  print("");

  print("The largest Jordan-Polya number less than 100 million: ${belowHundredMillion ?? 'Not found'}");
  print("");

  for (int i in [800, 1050, 1800, 2800, 3800]) {
    var decomposition = decompositions[jordanPolya[i - 1]];
    if (decomposition != null) {
      print("The ${i}th Jordan-Polya number is: ${jordanPolya[i - 1]} = ${mapToString(decomposition)}");
    }
  }
}

SplayTreeSet<int> jordanPolyaSet = SplayTreeSet<int>();
Map<int, Map<int, int>> decompositions = {};

void createJordanPolya() {
  jordanPolyaSet.add(1);
  Set<int> nextSet = SplayTreeSet<int>();
  decompositions[1] = {};
  int factorial = 1;

  for (int multiplier = 2; multiplier <= 20; multiplier++) {
    factorial *= multiplier;
    for (int number in jordanPolyaSet) {
      int tempNumber = number;
      while (tempNumber <= 9223372036854775807 ~/ factorial) {
        int original = tempNumber;
        tempNumber *= factorial;
        nextSet.add(tempNumber);
        var originalDecomposition = decompositions[original];
        if (originalDecomposition != null) {
          decompositions[tempNumber] = Map<int, int>.from(originalDecomposition)
            ..update(multiplier, (value) => value + 1, ifAbsent: () => 1);
        }
      }
    }
    jordanPolyaSet.addAll(nextSet);
    nextSet.clear();
  }
}

String mapToString(Map<int, int> map) {
  String result = "";
  map.forEach((key, value) {
    result = "$key!${value == 1 ? '' : '^$value'} * $result";
  });
  return result.isEmpty ? result : result.substring(0, result.length - 3);
}

File: Task/Jordan-P-lya-numbers/JavaScript/jordan-p-lya-numbers.js
class JordanPolyaNumbers {
    constructor() {
        this.jordanPolyaSet = new Set();
        this.decompositions = new Map();
    }

    main() {
        this.createJordanPolya();

        const jordanPolyaArray = Array.from(this.jordanPolyaSet).sort((a, b) => a - b);
        const belowHundredMillion = jordanPolyaArray.filter(n => n < 100000000).pop();

        console.log("The first 50 Jordan-Polya numbers:");
        let output = "";
        for (let i = 0; i < 50; i++) {
            output += jordanPolyaArray[i].toString().padStart(5, ' ');
            output += (i % 10 === 9 ? "\n" : "");
        }
        console.log(output);
        console.log();

        console.log(`The largest Jordan-Polya number less than 100 million: ${belowHundredMillion}`);
        console.log();

        for (const i of [800, 1050, 1800, 2800, 3800]) {
            const number = jordanPolyaArray[i - 1];
            console.log(`The ${i}th Jordan-Polya number is: ${number} = ${this.toString(this.decompositions.get(number))}`);
        }
    }

    createJordanPolya() {
        this.jordanPolyaSet.add(1);
        const nextSet = new Set();
        this.decompositions.set(1, new Map());
        let factorial = 1;

        for (let multiplier = 2; multiplier <= 20; multiplier++) {
            factorial *= multiplier;

            for (const num of this.jordanPolyaSet) {
                let number = num;
                while (number <= Number.MAX_SAFE_INTEGER / factorial) {
                    const original = number;
                    number *= factorial;
                    nextSet.add(number);

                    // Copy the decomposition from original
                    const newDecomp = new Map(this.decompositions.get(original));
                    this.decompositions.set(number, newDecomp);

                    // Merge: increment the count for this multiplier
                    const currentCount = newDecomp.get(multiplier) || 0;
                    newDecomp.set(multiplier, currentCount + 1);
                }
            }

            for (const num of nextSet) {
                this.jordanPolyaSet.add(num);
            }
            nextSet.clear();
        }
    }

    toString(map) {
        let result = "";
        const keys = Array.from(map.keys()).sort((a, b) => a - b);

        for (const key of keys) {
            const value = map.get(key);
            result = `${key}!${value === 1 ? '' : '^' + value} * ${result}`;
        }

        return result.substring(0, result.length - 3);
    }
}

// Run the program
const jp = new JordanPolyaNumbers();
jp.main();
